src/MongoDB.Driver.Core/Core/Authentication/ScramCache.cs

```diff
@@ -0,0 +1,114 @@
+/* Copyright 2018–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Linq;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// A cache for Client and Server keys, to be used during authentication.
+    /// </summary>
+    internal class ScramCache
+    {
+        private Object _cacheKey;
+        private Object _cacheValue;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        public ScramCacheEntry Get(Object key)
+        {
+            if (_cacheKey != null && _cacheKey.Equals(key))
+            {
+                return (ScramCacheEntry)_cacheValue;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        public void Set(Object key, Object value) //Tuple<string, string, int> key, byte[] clientKey, byte[] serverKey)
+        {
+            _cacheKey = key;
+            _cacheValue = value;
+        }
+    }
+
+    internal class ScramCacheKey
+    {
+        private int _iterationCount;
+        private byte[] _hashedPassword;
+        private byte[] _salt;
+
+        internal ScramCacheKey(byte[] hashedPassword, byte[] salt, int iterationCount)
+        {
+            _iterationCount = iterationCount;
+            _hashedPassword = hashedPassword;
+            _salt = salt;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (this == obj)
+            {
+                return true;
+            }
+
+            if (obj == null || obj.GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            ScramCacheKey other = (ScramCacheKey) obj;
+
+            return
+                _hashedPassword.SequenceEqual(other._hashedPassword) &&
+                _iterationCount == other._iterationCount &&
+                _salt.SequenceEqual(other._salt);
+        }
+
+        public override int GetHashCode()
+        {
+            int hash = 17;
+            hash = 37 * hash + _iterationCount.GetHashCode();
+            hash = 37 * hash + _hashedPassword.GetHashCode();
+            hash = 37 * hash + _salt.GetHashCode();
+            return hash;
+        }
```

Comment:
If we change the type of _password to SecureString you will have to figure out how to compute the hash code of a SecureString. Search and see if any other class has already solved that problem.

Reply:
The only case I've found in our code base is [this](https://github.com/ksadoff/mongo-csharp-driver/blob/490098591673fe534280a7bfdcf87942bcc184be/src/MongoDB.Driver/PasswordEvidence.cs#L90), but I'm not sure if it's transferable to our situation here.

Reply:
So computing the hash code of a SecureString is expensive. Let's just ignore it.

The following GetHashCode implementation follows our normal standards for GetHashCode:

    public override int GetHashCode()
    {
        // ignore _password when computing the hash code
        return new Hasher()
            .Hash(_iterationCount)
            .Hash(_salt)
            .GetHashCode();
    }

Why is it OK to ignore _password? Well remember that even the following implementation is legal:

    public override int GetHashCode() => 0;

Reply:
I've included this implementation of GetHashCode.